rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -151,64 +164,70 @@ class ReplayRSpace[F[_]: Sync, C, P, A, K](
               }
       } yield r
 
-    def getCommOrDataCandidates(comms: Seq[COMM]): F[Either[COMM, Seq[DataCandidate[C, A]]]] = {
-      type COMMOrData = Either[COMM, Seq[DataCandidate[C, A]]]
-      def go(comms: Seq[COMM]): F[Either[Seq[COMM], Either[COMM, Seq[DataCandidate[C, A]]]]] =
+    def getCommAndDataCandidates(
+        comms: Seq[COMM]
+    ): F[Option[(COMM, Seq[DataCandidate[C, A]])]] = {
+      type COMMOrData = Either[COMM, (COMM, Seq[DataCandidate[C, A]])]
```

Comment:
Maybe it would make sense to have the `(COMM, Seq[DataCandidate[C, A]])` also named as a `type`? all this deserves a through refactor IMO...

Reply:
Agreed, there is a lot of refactoring potential here